### PR TITLE
feat: add confirmation dialog before deletion

### PR DESCRIPTION
**Pull request by NilinReza**
- Add delete button to delete-recipe.component
- Add check in deleteRecipe to assert confirmation from admin
- Add new Dialog service to handle confirmation
- Add Dialog component and Dialog interface

### CODE REVIEW

File: client/src/services/dialog/dialog.service.ts [hunk 1]
```diff
@@ -0,0 +1,41 @@
...
+    secondaryBtnText: '',
+  };
+
+  constructor(private dialog: MatDialog) {}
+
+  public get dialog$() {
+    return this._dialog$.asObservable();
+  }
+
+  public confirmDelete(recipeID: string): Observable<boolean> {
```
Comment (zaidi-01) [inline client/src/services/dialog/dialog.service.ts:25]: This service needs to be generic as well. Rename the method to `open`, similar to `MatDialog`. Then construct dialog data in the recipe id input component and pass it here.

File: client/src/interfaces/dialog.interface.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
...
+  /** The confirmation message. */
+  message: string;
+
+  recipeID: string
+  /** The ID of the recipe to be deleted. */
+  primaryBtnText: string;
+
+  secondaryBtnText?: string;
+
+}
```
Comment (zaidi-01) [inline client/src/interfaces/dialog.interface.ts:17]: Introduce a `type` for the dialog.

File: client/src/services/dialog/dialog.component.html [hunk 3]
```diff
@@ -0,0 +1,19 @@
+<div>
+  <form class="modal-content container">
```
Comment (zaidi-01) [inline client/src/services/dialog/dialog.component.html:2]: Why a form?

File: client/src/services/dialog/dialog.component.html [hunk 4]
```diff
@@ -0,0 +1,19 @@
...
+    <span class="close" [mat-dialog-close]="false">&times;</span>
+    <mat-dialog-content class="mat-typography">
+      <h2 mat-dialog-title class="dialog-title">{{ data.title }}</h2>
+      <p class="dialog-message">{{ data.message }}</p>
+      <b> Deletion is irreversible and may cause unwanted behavior</b>
+    </mat-dialog-content>
+    <hr />
+    <mat-dialog-actions class="clearfix">
+      <button mat-button class="cancelbtn" [mat-dialog-close]="false">
```
Comment (zaidi-01) [inline client/src/services/dialog/dialog.component.html:11]: Script this according to the `type` of the dialog. For example, only show the cancel button if the dialog is of type `Confirmation`.

**Review by zaidi-01 — APPROVED** [review]

File: client/src/admin/delete-recipe/delete-recipe.component.ts [hunk 5]
```diff
@@ -11,20 +12,48 @@ import { Observable, catchError, tap } from 'rxjs';
...
           if (err instanceof HttpErrorResponse && err.status === 404) {
             alertService.error(`Recipe #${recipeId} not found`);
           } else {
             alertService.error(`Failed to delete recipe #${recipeId}`);
           }
           throw err;
-        })
+        }),
+
+        map(() => {})
```
[the remark this answers is not in the document]
Comment (NilinReza) [reply client/src/admin/delete-recipe/delete-recipe.component.ts:56]: The `map` on line 56? The `deleteRecipe` method in delete-recipe.component is supposed to be returning `Observable<void>` so removing that line will mean that I am returning `void` only

File: client/src/admin/delete-recipe/delete-recipe.component.ts [hunk 6]
```diff
@@ -11,20 +12,48 @@ import { Observable, catchError, tap } from 'rxjs';
...
+    alertService: AlertService,
+    private dialogService: DialogService
+  ) {
+    this.deleteRecipe = (recipeId: string): Observable<void> => {
       alertService.clear();
 
-      return recipeService.deleteRecipe(+recipeId).pipe(
-        tap(() => alertService.success(`Recipe #${recipeId} deleted`)),
+      // First, get the recipe
+      return from(recipeService.getRecipe(+recipeId).toPromise()).pipe(
```
[the remark this answers is not in the document]
Comment (NilinReza) [reply client/src/admin/delete-recipe/delete-recipe.component.ts:24]: Noted

**Review by JJTomines — APPROVED** [review]

**Pull request merged**